### PR TITLE
[JENKINS-61920] Avoid MissingClassTelemetry populates the log

### PR DESCRIPTION
**Pull request by MRamonLeon**
See [JENKINS-61920](https://issues.jenkins-ci.org/browse/JENKINS-61920).

### Proposed changelog entries

* Internal: Avoid populating logs with missing classes telemetry warnings

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgade
- [x] Appropriate autotests or explanation to why this change has no tests
- [N/A] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@jenkinsci/core-pr-reviewers 

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [x] There are at least 2 approvals for the pull request and no outstanding requests for change
- [x] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [x] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [x] Proper changelog labels are set so that the changelog can be generated automatically
- [x] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [x] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]

File: core/src/main/java/jenkins/telemetry/impl/java11/MissingClassEvents.java [hunk 1]
```diff
@@ -98,6 +109,22 @@ public long put(String name, @NonNull Throwable t) {
         return currentEvents;
     }
 
+    /**
+     * Returns true if the class name was already registered previously, during the current execution of this Jenkins
+     * instance.
+     * @ param t the throwable to check
```
Comment (varyvol) [inline core/src/main/java/jenkins/telemetry/impl/java11/MissingClassEvents.java:115]: Can be removed.

File: core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java [hunk 2]
```diff
@@ -239,13 +241,18 @@ public static void reportException(@NonNull String name, @NonNull Throwable e) {
             // We call the methods in this order because if the missing class is not java related, we don't loop over the
             // stack trace to look if it's not thrown from an ignored place avoiding an impact on performance.
             if (isFromMovedPackage(name) && !calledFromIgnoredPlace(e)) {
-                events.put(name, e);
-                if (LOGGER.isLoggable(Level.WARNING))
+                if (LOGGER.isLoggable(Level.WARNING) && !wasAlreadyReported(name)) {
                     LOGGER.log(Level.WARNING, "Added a missed class for missing class telemetry. Class: " + name, e);
+                }
+                events.put(name, e);
```
Comment (varyvol) [inline core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java:247]: Should it be added even if already there?

File: core/src/main/java/jenkins/telemetry/impl/java11/MissingClassEvents.java [hunk 3]
```diff
@@ -25,8 +25,8 @@
 package jenkins.telemetry.impl.java11;
 
 import com.google.common.annotations.VisibleForTesting;
-
 import edu.umd.cs.findbugs.annotations.NonNull;
+
```
Comment (jglick) [inline core/src/main/java/jenkins/telemetry/impl/java11/MissingClassEvents.java:29]: Please avoid diff hunks unrelated to a fix.

File: core/src/main/java/jenkins/telemetry/impl/java11/MissingClassEvents.java [hunk 4]
```diff
@@ -42,10 +42,15 @@
...
 
     /**
-     * List of events, one per stack trace.
+     * List of events, one per stack trace, to send to Telemetry. Only the first {@code #MAX_EVENTS_PER_SEND} are sent.
      */
     private ConcurrentHashMap<List<StackTraceElement>, MissingClassEvent> events = new ConcurrentHashMap<>(MAX_EVENTS_PER_SEND);
 
+    /**
+     * List of all events registered on this execution, to avoid printing an event more than once in the log. This map
+     * is not limited. On every Jenkins restart is cleaned because it's not persisted, so the CNFE is printed again.
```
Comment (jglick) [inline core/src/main/java/jenkins/telemetry/impl/java11/MissingClassEvents.java:51]: Best to document what the key represents.

File: core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java [hunk 5]
```diff
@@ -107,7 +107,9 @@
             {"org.codehaus.groovy.control.ClassNodeResolver", "tryAsLoaderClassOrScript"},
             {"org.kohsuke.stapler.RequestImpl$TypePair", "convertJSON"},
             {"net.bull.javamelody.FilterContext", "isMojarraAvailable"}, // JENKINS-60725
-            {"hudson.remoting.RemoteClassLoader$ClassLoaderProxy", "fetch3"} // JENKINS-61521
+            {"hudson.remoting.RemoteClassLoader$ClassLoaderProxy", "fetch3"}, // JENKINS-61521
+            //Don't add "java.base/" before sun.reflect.generics.factory.CoreReflectionFactory
+            {"sun.reflect.generics.factory.CoreReflectionFactory", "makeNamedType"} // JENKINS-61920
```
Comment (jglick) [inline core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java:112]:
```suggestion
            {"sun.reflect.generics.factory.CoreReflectionFactory", "makeNamedType"}, // JENKINS-61920
```

so you do not need to patch the previous line the next time you add an entry

**Review by jglick — COMMENTED** [review]
Do not understand code well enough to really review.

File: core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java (hunk 5 above)
Comment (MRamonLeon) [reply core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java:112]: I've done it on other languages, I never tried it on Java :smile:

File: core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java (hunk 2 above)
Comment (MRamonLeon) [reply core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java:247]: The put method is not the Map one, is a custom method which makes two things:

- Put the event in the list of already printed, if needed
- Put the event in the list of events to send to Telemetry, if there is space for it (only 100 per day)

**Review by MRamonLeon — COMMENTED** [review]
Addressed feedbacks

**Review by varyvol — APPROVED** [review]

**Comment by timja** [comment]
This is ready for merge now, we'll merge it in 24 hours if no negative feedback

**Pull request merged**